### PR TITLE
[Doc] Refactor field

### PR DESCRIPTION
**Pull request by neozhaoliang**

### CODE REVIEW

**Review by Olinaaaloompa — CHANGES_REQUESTED** [review]

File: docs/lang/articles/basic/field.md [hunk 1]
```diff
@@ -239,37 +419,45 @@ The following code snippet declares a 1D field of particle information (position
...
-particle_field[1].pos[0] = 1.0 # pos[0] is the first member of pos
++ By the name-to-index way, as shown in the code below, you first access the field which gathers all the `mass` members in the struct field, and then use the index to access a specific one:
 
-# Universally set the mass of all particles to 1.0
-particle_field.mass.fill(1.0)
+```python
+particle_field.mass[0] = 1.0  # Set the mass of the first particle in the field to 1.0
+```
+
+Since `paticle_field.mass` is a field that consists of all the `mass` members of the structs in `paticle_field`, we can also call its `fill()` method to set them to a specific value all at once:
```
Comment (Olinaaaloompa) [inline docs/lang/articles/basic/field.md:459]:
```suggestion
Considering that `paticle_field.mass` is a field consisting of all the `mass` members of the structs in `paticle_field`, we can also call its `fill()` method to set the members to a specific value all at once:
```

**Review by Olinaaaloompa — CHANGES_REQUESTED** [review]

File: docs/lang/articles/basic/field.md [hunk 2]
```diff
@@ -120,18 +286,30 @@ Accessing a vector field is similar to accessing a multi-dimensional array: You
...
 - To access the velocity vector at a specific position of the volumetric field above:
 
-  `volumetric_field[i,j,k]`
+  `volumetric_field[i, j, k]`
 
 - To access the `l`-th component of the velocity vector:
 
   `volumetric_field[i, j, k][l]`
 
+- If the vectors in the field have dimension <=4, then you can also use vector swizzling indices `xyzw` or `rgba` to access the components of the vectors:
```
[the remark this answers is not in the document]
Comment (neozhaoliang) [reply docs/lang/articles/basic/field.md:298]: "provided that the vector has no more than four dimensions:"

It should be "provided that the dimension of the vector is no more than four".

**Review by Olinaaaloompa — APPROVED** [review]

File: docs/lang/articles/basic/field.md [hunk 3]
```diff
@@ -14,42 +14,170 @@ A 0D (zero-dimensional) field contains *only* one element.
...
-# Declare a 2D scalar field whose shape is (640, 480)
-f_2d = ti.field(ti.u8,  shape=(640, 480))   # 2D
-# Declare a 3D scalar field whose data type is f32
-f_3d = ti.field(ti.f32, shape=(32, 32, 32)) # 3D
-```
 
 ### Access elements in a scalar field
 
-The initial value of elements in a scalar filed is zero. Always use explicit indexing to access elements in a scalar field.
+Once a field is declared, Taichi automatically assigns an initial value of zero to its elements. But we do not recommend using this undefined behavior.
```
Comment (yuanming-hu) [inline docs/lang/articles/basic/field.md:120]:
```suggestion
Once a field is declared, Taichi automatically assigns an initial value of zero to its elements.
```

Zero-initialization is defined behavior.

File: docs/lang/articles/basic/field.md [hunk 4]
```diff
@@ -14,42 +14,170 @@ A 0D (zero-dimensional) field contains *only* one element.
...
+                │  │   │       │   │  │
+                └  └───┴╌╌╌╌╌╌╌┴───┘  ┘
+
+      ┌───┬───┬───┬───┬───┬───┬───┬───┐
+  * = │   │   │   │   │   │   │   │   │
+      └───┴───┴───┴───┴───┴───┴───┴───┘
+      └───────────────────────────────┘
+             f_1d.shape[2] = 8
+
+  ```
```
Comment (yuanming-hu) [inline docs/lang/articles/basic/field.md:107]: I feel like this is too verbose. The readers should understand what a high-dimensional array is. Let's remove the 3D case here.

File: docs/lang/articles/basic/field.md [hunk 5]
```diff
@@ -14,42 +14,170 @@ A 0D (zero-dimensional) field contains *only* one element.
...
+            f_1d.shape = (9,)
+    ```
+
++ To declare a 2D scalar field, set the size of its first two dimensions, i.e. the number of rows and columns, respectively. For example, the following code defines a 2D scalar field of shape (3, 6), which has 3 rows and 6 columns:
+
+    ```python
+    f_2d = ti.field(int, shape=(3, 6))  # 2D field
+    ```
+
+    The memory layout of the field `f_2d` will be like:
```
Comment (yuanming-hu) [inline docs/lang/articles/basic/field.md:63]:
```suggestion
    An illustration of f_2d is below:
```
Let's avoid using `memory layout` here, which has a different meaning. Just use `An illustration of f_2d is below` is good enough.

File: docs/lang/articles/basic/field.md [hunk 6]
```diff
@@ -73,37 +201,78 @@ while gui.running:
...
+For example, the following code snippet declares a 2D field of 2D vectors:
 
 ```python
-# Declare a 1x2x3 vector field, whose vector dimension is n=2
-f = ti.Vector(ti.f32, n=2).field(shape=(1,2,3))
+# Declare a 3x3 vector field comprising 2D vectors
+f = ti.Vector.field(n=2, dtype=float, shape=(3, 3))
+```
+
+The memory layout of `f` will be like:
```
Comment (yuanming-hu) [inline docs/lang/articles/basic/field.md:262]: Avoid the term `memory layout`.

File: docs/lang/articles/basic/field.md [hunk 7]
```diff
@@ -120,18 +289,30 @@ Accessing a vector field is similar to accessing a multi-dimensional array: You
...
 
+- Alternatively, you can use swizzling with the indices `xyzw` or `rgba` to access the components of a vector, provided that the dimension of the vector is no more than four:
+
+  ```python
+  volumetric_field[i, j, k].x = 1  # equivalent to volumetric_field[i, j, k][0] = 1
+  volumetric_field[i, j, k].y = 2  # equivalent to volumetric_field[i, j, k][1] = 2
+  volumetric_field[i, j, k].z = 3  # equivalent to volumetric_field[i, j, k][2] = 3
+  volumetric_field[i, j, k].w = 4  # equivalent to volumetric_field[i, j, k][3] = 4
+  volumetric_field[i, j, k].xyz = 1, 2, 3  # assign 1, 2, 3 to the first three components
+  volumetric_field[i, j, k].rgb = 1, 2, 3  # ibid.
```
Comment (yuanming-hu) [inline docs/lang/articles/basic/field.md:306]: Avoid `ibid`. Not clear what that means.

**Review by yuanming-hu — COMMENTED** [review]
A few comments.

File: docs/lang/articles/basic/field.md [hunk 8]
```diff
@@ -14,42 +14,140 @@ A 0D (zero-dimensional) field contains *only* one element.
...
+     f_0d.shape=()
+    ```
+
++ To declare a 1D scalar field of length `n`, set its shape to `n` or `(n,)`:
+
+    ```python
+    f_1d = ti.field(ti.i32, shape=9)  # 1D field
+    ```
+
+    An illustration of `f_1d` shown is below:
```
Comment (Olinaaaloompa) [inline docs/lang/articles/basic/field.md:47]:
```suggestion
    An illustration of `f_1d` is shown below:
```

File: docs/lang/articles/basic/field.md [hunk 9]
```diff
@@ -14,42 +14,140 @@ A 0D (zero-dimensional) field contains *only* one element.
...
 
 
 ### Declaration
 
-```python
-import taichi as ti
-ti.init(arch=ti.cpu)
+The simplest way to declare a scalar field is to call `ti.field(dtype, shape)`, where `dtype` is a primitive data type as explained in [type system](../type_system/type.md) and `shape` is a tuple of integers.
+
++ To declare a 0D scalar field, set its shape to the empty tuple `()`.
```
Comment (Olinaaaloompa) [inline docs/lang/articles/basic/field.md:24]:
```suggestion
+ To declare a 0D scalar field, set its shape to the empty tuple `()`:
```

File: docs/lang/articles/basic/field.md [hunk 10]
```diff
@@ -14,42 +14,140 @@ A 0D (zero-dimensional) field contains *only* one element.
...
+
+    ```
+    ┌───┬───┬───┬───┬───┬───┬───┬───┬───┐
+    │   │   │   │   │   │   │   │   │   │
+    └───┴───┴───┴───┴───┴───┴───┴───┴───┘
+    └───────────────────────────────────┘
+            f_1d.shape = (9,)
+    ```
+
++ To declare a 2D scalar field, set the size of its first two dimensions, i.e. the number of rows and columns, respectively. For example, the following code defines a 2D scalar field of shape (3, 6), which has 3 rows and 6 columns:
```
Comment (Olinaaaloompa) [inline docs/lang/articles/basic/field.md:57]:
```suggestion
+ To declare a 2D scalar field, set the sizes of its first two dimensions, i.e., the number of rows and columns, respectively. For example, the following code defines a 2D scalar field of shape (3, 6), which has 3 rows and 6 columns:
```

File: docs/lang/articles/basic/field.md [hunk 11]
```diff
@@ -239,37 +390,45 @@ The following code snippet declares a 1D field of particle information (position
...
+  pos=vec3, vel=vec3, acc=vec3, mass=float,
 )
 # Declare a 1D field of the struct particle using field()
 particle_field = particle.field(shape=(n,))
 ```
 
 ### Access elements in a struct field
 
-You can access members of elements in a struct field either one by one or universally:
+You can access a member of an element in a struct field in either of the following ways: "index-first"  or "name-first".
```
Comment (Olinaaaloompa) [inline docs/lang/articles/basic/field.md:415]:
```suggestion
You can access a member of an element in a struct field in either of the following ways: "index-first" or "name-first".
```

**Review by Olinaaaloompa — CHANGES_REQUESTED** [review]

**Review by Olinaaaloompa — APPROVED** [review]

**Pull request merged**